src/surface/polygon_soup_mesh.cpp

```diff
@@ -120,6 +148,177 @@ void PolygonSoupMesh::readMeshFromFile(std::string filename) {
+    }
+
+    polygons.push_back(face);
+  }
+}
+
+void PolygonSoupMesh::readMeshFromBinaryStlFile(std::ifstream in) {
+  auto parseVector3 = [&](std::ifstream& in) {
+    char buffer[3 * sizeof(float)];
+    in.read(buffer, 3 * sizeof(float));
+    float* fVec = (float*)buffer;
+    return Vector3{fVec[0], fVec[1], fVec[2]};
+  };
+
+  char header[80];
+  char nTriangleChars[4];
+  in.read(header, 80);
+  in.read(nTriangleChars, 4);
+  unsigned int* intPtr = (unsigned int*)nTriangleChars;
+  size_t nTriangles = *intPtr;
+
+  for (size_t iT = 0; iT < nTriangles; ++iT) {
+    // TODO: store this normal?
+    Vector3 normal = parseVector3(in);
+    std::vector<size_t> face;
+    for (size_t iV = 0; iV < 3; ++iV) {
+      vertexCoordinates.push_back(parseVector3(in));
+      face.push_back(vertexCoordinates.size() - 1);
+    }
+
+    // Orient face using normal
+    Vector3 faceNormal = cross(vertexCoordinates[face[1]] - vertexCoordinates[face[0]],
+                               vertexCoordinates[face[2]] - vertexCoordinates[face[0]]);
+    if (dot(faceNormal, normal) < 0) {
+      std::reverse(std::begin(face), std::end(face));
+    }
+
+    polygons.push_back(face);
+    char dummy[2];
+    in.read(dummy, 2);
+  }
+}
+
+// Read a .stl file containing a polygon mesh
+void PolygonSoupMesh::readMeshFromStlFile(std::string filename) {
+  // TODO: read stl file name
+  polygons.clear();
+  vertexCoordinates.clear();
+
+  // Open the file
+  std::ifstream in(filename);
+  if (!in) throw std::invalid_argument("Could not open mesh file " + filename);
+
+  // parse stl format
+  std::string line;
+  getline(in, line);
+  std::stringstream ss(line);
+  std::string token;
+  ss >> token;
+  if (token == "solid") {
+    readMeshFromAsciiStlFile(in);
+  } else {
+    readMeshFromBinaryStlFile(std::ifstream(filename, std::ios::in | std::ios::binary));
+  }
+}
+
+// Mutate this mesh by merging vertices with identical floating point positions.
+// Useful for loading .stl files, which don't contain information about which
+// triangle corners meet at vertices.
+void PolygonSoupMesh::mergeIdenticalVertices() {
+  std::vector<Vector3> compressedPositions;
+  // Store mapping from original vertex index to merged vertex index
+  std::vector<size_t> compressVertex;
+  compressVertex.reserve(vertexCoordinates.size());
+
+  std::unordered_map<Vector3, size_t> canonicalIndex;
+
+  for (size_t iV = 0; iV < vertexCoordinates.size(); ++iV) {
+    Vector3 v = vertexCoordinates[iV];
+    auto it = canonicalIndex.find(v);
+
+    // Check if vertex exists in map or not
+    if (it == canonicalIndex.end()) {
+      compressedPositions.push_back(v);
+      size_t vecIndex = compressedPositions.size() - 1;
+      canonicalIndex[v] = vecIndex;
+      compressVertex.push_back(vecIndex);
+    } else {
+      size_t vecIndex = it->second;
+      compressVertex.push_back(vecIndex);
+    }
+  }
+
+  vertexCoordinates = std::move(compressedPositions);
+  compressedPositions.clear();
```

Comment:
Calling a method on `compressedPositions` after moving from it is undefined behavior ("use after move"), right? I think this line should be omitted.

Reply:
You're right. I misunderstood an answer on StackExchange. I've deleted the `clear()`.